pkg/cachemanager/cachemanager.go

```diff
@@ -140,47 +156,82 @@ func (c *CacheManager) UpsertSource(ctx context.Context, sourceKey aggregator.Ke
 func (c *CacheManager) replaceWatchSet(ctx context.Context) error {
 	newWatchSet := watch.NewSet()
 	newWatchSet.Add(c.gvksToSync.GVKs()...)
+	c.gvksToDeleteFromCache.AddSet(c.watchedSet.Difference(newWatchSet))
 
-	diff := c.watchedSet.Difference(newWatchSet)
-	c.removeStaleExpectations(diff)
-
-	c.gvksToDeleteFromCache.AddSet(diff)
-
-	var innerError error
+	var err error
 	c.watchedSet.Replace(newWatchSet, func() {
 		// *Note the following steps are not transactional with respect to admission control
 
 		// Important: dynamic watches update must happen *after* updating our watchSet.
 		// Otherwise, the sync controller will drop events for the newly watched kinds.
-		innerError = c.registrar.ReplaceWatch(ctx, newWatchSet.Items())
+		err = c.registrar.ReplaceWatch(ctx, newWatchSet.Items())
 	})
 
-	return innerError
+	if err != nil {
+		// account for any watches failing to remove
+		if f := watch.NewErrorList(); errors.As(err, &f) {
+			c.handleDanglingWatches(f.RemoveGVKFailures())
```

Comment:
It doesn't look like this is handling the case of a general error?

Reply:
Thanks for the call out -- I think the general case should be handled now 🙏🏼